core/util/src/main/java/org/eclipse/rdf4j/common/iteration/UnionIteration.java

```diff
@@ -62,24 +62,26 @@ protected E getNextElement() throws X {
 		if (isClosed()) {
 			return null;
 		}
-		Iteration<? extends E, X> nextCurrentIter = currentIter;
-		if (nextCurrentIter != null && nextCurrentIter.hasNext()) {
-			return nextCurrentIter.next();
-		}
 
-		// Current Iteration exhausted, continue with the next one
-		Iterations.closeCloseable(nextCurrentIter);
+		while (true) {
```

Comment:
I don't particularly like the notion of an infinite loop. Is there a way we can move the conditional to jump out of the loop in here? `while(argIter.hasNext())` perhaps? I'm not sure how that would play out with the synchronized block to be honest (not even sure why we're synchronizing inside an iterator etiher, tbh, but that's a separate issue).

Reply:
It's hard to make sure all the conditions are met correctly, especially with that synchronized block. Not sure why there is a synchronized block to be honest...is the code assuming that there will be some sort of parallel access to this iteration?

Don't like the while(true) either, but it's essentially what was there already.

Reply:
Fair enough. The synchronized block might be a leftover from the past, when there was a bit of a misguided "better safe than sorry" philosophy towards concurrency. I'm not convinced it's useful, but ripping it out here might be a bit beyond the scope of your intended fix.